### PR TITLE
mosquitto.sh runs native mosquitto

### PR DESCRIPTION
**Pull request by jpgrayson**
I am posting this as a possible solution to the problem of how to run tests natively with a mosquitto service.

This changes `scripts/mosquitto.sh` to run a native mosquitto process instead of running mosquitto as a containerized service. This specifically allows mosquitto to create/bind a unix socket path that is available to natively running tests.

This is the best way forward I am seeing to enable tests running natively on MacOS to support unix sockets. But I also appreciate that it's a very different paradigm.

The container based approach works on Linux; i.e. a container can share a unix socket to a Linux host. So `mosquitto.sh` *could* be written to use containers on Linux and only use a native mosquitto process on MacOS. Let me know if that alternative seems right.

### CODE REVIEW

**Comment by adam-fowler** [comment]
Is it not possible to have mosquitto bind a path internal to the docker container?

**Comment by jpgrayson** [comment]
Mosquitto can bind a unix socket path internal to the docker container, but tests running in the host MacOS environment need access to that socket path.

Using a volume mount is the nominal way to share parts of the host and container filesystems, but a unix socket is not a regular file and cannot be shared between a Linux container and a MacOS host.

A unix socket can be shared between a container and a linux host.

AFAICT, the only way to enable tests running natively on MacOS to connect to mosquitto via a unix socket will involve running a native mosquitto process on the MacOS host.

This change modifies `mosquitto.sh` for this purpose.

As noted, we could structure this script to continue using the containerized mosquitto approach on Linux hosts where that will work with unix sockets.

**Comment by adam-fowler** [comment]
This script was originally written on the understanding mosquitto was running inside Docker and might be running in the background. We can probably remove that and just support running in the foreground now.

**Comment by jpgrayson** [comment]
I can make the script just run in the foreground. That would be simpler and would avoid monkeying with a pid file.

I'd kind of like to have the option on Linux of running either non-containerized mosquitto or in a container. Perhaps by testing for a `mosquitto` executable and falling back to running a container.

I'll update the PR with these changes. Let me know what you think. And thanks again for taking a look at these changes.

**Comment by adam-fowler** [comment]
@jpgrayson this all sounds good.

File: scripts/mosquitto.sh [hunk 1]
```diff
@@ -1,83 +1,58 @@
...
-        echo "Mosquitto is already running"
-    fi
+    docker run \
+        -p 1883:1883 \
+        -p 1884:1884 \
+        -p 8883:8883 \
+        -p 8080:8080 \
+        -p 8081:8081 \
+        -v ./mosquitto/config:/mosquitto/config \
+        -v ./mosquitto/certs:/mosquitto/certs \
```
Comment (adam-fowler) [inline scripts/mosquitto.sh:34]: These folder mappings need a full path. It fails otherwise

File: scripts/mosquitto.sh [hunk 2]
```diff
@@ -1,83 +1,58 @@
...
 fi
 
+cd "$(dirname "$(dirname "$0")")"
+
+if [[ $USE_CONTAINER -eq 1 ]]; then
+    run-containerized-mosquitto
+elif command -v mosquitto >/dev/null; then
+    run-installed-mosquitto
+else
+    echo "notice: mosquitto not installed; running eclipse-mosquitto container instead..."
```
Comment (adam-fowler) [inline scripts/mosquitto.sh:56]: Maybe add how to install mosquitto

File: scripts/mosquitto.sh (hunk 1 above)
Comment (jpgrayson) [reply scripts/mosquitto.sh]: Interesting. The relative paths worked for me, but I'm using podman, not docker.

I changed this to absolute paths in the latest version of the change.

File: scripts/mosquitto.sh (hunk 2 above)
Comment (jpgrayson) [reply scripts/mosquitto.sh:59]: Done

**Review by adam-fowler — APPROVED** [review]

**Pull request merged**